File: apps/gooseforum/resource/src/site/utils/pkCourseOrder.ts
type CourseCodeItem = { courseCode: string }

/** 稳定地将满足计划条件的项目置顶，保留两组内的原始顺序。 */
export function sortPlannedFirst<T>(items: readonly T[], isPlanned: (item: T) => boolean): T[] {
  const planned: T[] = []
  const other: T[] = []

  for (const item of items) {
    const bucket = isPlanned(item) ? planned : other
    bucket.push(item)
  }

  return [...planned, ...other]
}

/** 将专业计划内课程置顶，同时保留两组内的接口/备选池顺序。 */
export function sortPlannedCoursesFirst<T extends CourseCodeItem>(
  courses: readonly T[],
  compulsoryCourses: readonly CourseCodeItem[],
): T[] {
  const plannedCodes = new Set(compulsoryCourses.map((course) => course.courseCode))
  return sortPlannedFirst(courses, (course) => plannedCodes.has(course.courseCode))
}

File: apps/mobile/packages/core/lib/src/schedule/pk_course_order.dart
/// 排课列表排序：专业计划内课程置顶（移植自 web `pkCourseOrder.ts`）。
library;

/// 将专业计划内课程置顶，同时保留两组内的接口/备选池顺序（稳定分区）。
List<T> sortPlannedCoursesFirst<T>(
  List<T> courses,
  List<T> compulsoryCourses,
  String Function(T) courseCodeOf,
) {
  final plannedCodes = compulsoryCourses.map(courseCodeOf).toSet();
  final planned = <T>[];
  final other = <T>[];
  for (final course in courses) {
    (plannedCodes.contains(courseCodeOf(course)) ? planned : other).add(course);
  }
  return [...planned, ...other];
}
